> The present file is about using some specified strategies to clean the "Global Electricity Statistics" data. Each step has been completely explained.







*   In the first step we need to import the required libraries.



In [115]:
#Import libraries
import numpy as np
import pandas as pd



*   In order to use the dataset, we need to upload it.



In [133]:
#Upload the dataset from computer directory
from google.colab import files
uploaded = files.upload()

Saving Global Electricity Statistics.csv to Global Electricity Statistics (8).csv




*   The file is uploaded but we have to call it.


In [134]:
#Import data
data = pd.read_csv("Global Electricity Statistics.csv")



*   The next step is to explore the data and get familiar to its features.



In [135]:
#Find the dimensions of the dataset
data.shape

(1610, 45)

In [21]:
#First 5 rows of the dataset
data.head()

Country        Features  Region   1980   1981   1982   1983  \
0               Algeria  net generation  Africa  6.683   7.65  8.824  9.615   
1                Angola  net generation  Africa  0.905  0.906  0.995  1.028   
2                 Benin  net generation  Africa  0.005  0.005  0.005  0.005   
3              Botswana  net generation  Africa  0.443  0.502  0.489  0.434   
4          Burkina Faso  net generation  Africa  0.098  0.108  0.115  0.117   

     1984    1985    1986  ...     2012     2013      2014      2015  \
0  10.537  11.569  12.214  ...  53.9845  56.3134  60.39972  64.68244   
1   1.028   1.028   1.088  ...  6.03408  7.97606   9.21666   9.30914   
2   0.005   0.005   0.005  ...  0.04612  0.08848   0.22666   0.31056   
3   0.445   0.456   0.538  ...     0.33  0.86868   2.17628   2.79104   
4   0.113   0.115   0.122  ...  0.86834  0.98268   1.11808   1.43986   

        2016      2017      2018     2019         2020         2021  
0   66.75504  71.49546  72.10903   76.685  72.73591277  77.53072719  
1  10.203511  10.67604  12.83194     15.4         16.6    16.429392  
2    0.26004    0.3115   0.19028   0.2017      0.22608   0.24109728  
3    2.52984    2.8438   2.97076   3.0469      2.05144   2.18234816  
4     1.5509   1.64602    1.6464  1.72552  1.647133174  1.761209666  

[5 rows x 45 columns]

In [22]:
#The last 5 rows of the dataset
data.tail()

Country              Features  \
1605               Trinidad and Tobago  distribution losses    
1606          Turks and Caicos Islands  distribution losses    
1607               U.S. Virgin Islands  distribution losses    
1608                           Uruguay  distribution losses    
1609                         Venezuela  distribution losses    

                       Region     1980     1981     1982     1983     1984  \
1605  Central & South America    0.244     0.21    0.152    0.326     0.36   
1606  Central & South America  0.00035  0.00035  0.00035  0.00035  0.00035   
1607  Central & South America  0.05243  0.05537  0.05607  0.05761  0.05789   
1608  Central & South America     0.55    0.426    0.627    0.662     0.78   
1609  Central & South America    5.145    3.597    6.113    6.623    5.386   

         1985     1986  ...    2012    2013    2014    2015    2016  \
1605    0.407    0.337  ...   0.239   0.234   0.245   0.248   0.253   
1606  0.00035  0.00035  ...       0       0       0       0       0   
1607  0.05922  0.06055  ...   0.063    0.06    0.06    0.06   0.065   
1608    0.702    0.911  ...   1.292   1.282   1.253    1.49   1.608   
1609    6.818    8.605  ...  25.113  25.627  42.699  40.381  33.543   

           2017       2018       2019       2020         2021  
1605   0.274864   0.425807   0.424101   0.422757     0.422757  
1606      0.014     0.0125     0.0125     0.0125   0.01277172  
1607      0.065       0.05      0.051      0.051        0.051  
1608   1.694491   1.557257   1.322331   1.129273     1.129273  
1609  32.237348  30.982821  25.637959  23.099784  23.04704156  

[5 rows x 45 columns]

In [7]:
#Check basic information about the dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1610 entries, 0 to 1609
Data columns (total 45 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Country   1610 non-null   object
 1   Features  1610 non-null   object
 2   Region    1610 non-null   object
 3   1980      1582 non-null   object
 4   1981      1582 non-null   object
 5   1982      1582 non-null   object
 6   1983      1582 non-null   object
 7   1984      1582 non-null   object
 8   1985      1582 non-null   object
 9   1986      1582 non-null   object
 10  1987      1582 non-null   object
 11  1988      1582 non-null   object
 12  1989      1582 non-null   object
 13  1990      1582 non-null   object
 14  1991      1582 non-null   object
 15  1992      1582 non-null   object
 16  1993      1582 non-null   object
 17  1994      1582 non-null   object
 18  1995      1582 non-null   object
 19  1996      1582 non-null   object
 20  1997      1582 non-null   object
 21  1998      1582



* By looking at the above result, it is obvious that the data type in not defined correctly. The present data is a time series and its first columns include country, feature and region. In the same way, we include the datatype.



In [136]:
#Convert 'column1' to category
data['Country'] = data['Country'].astype('category')

#Convert 'column2' to category
data['Features'] = data['Features'].astype('category')

#Convert 'column3' to category
data['Region'] = data['Region'].astype('category')

#Convert 'column4' to 'column45' to float
data.iloc[:,3:45] = data.iloc[:,3:45].apply(pd.to_numeric, errors='coerce')

<ipython-input-136-352bf419a8e1>:11: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  data.iloc[:,3:45] = data.iloc[:,3:45].apply(pd.to_numeric, errors='coerce')


> The appropriate data type for the "country" column depends on the nature of the data it contains. Because the "country" column represents discrete categories (i.e., the names of countries), it is often best to treat it as a categorical variable. This is especially useful when you have a fixed and known set of countries.

> Because the dataset contains decimal numbers in a time series where there is no date information, we may want to represent the time axis using numeric values. The appropriate data type for the time axis would be a numeric type, such as "float" or "int". "float" is chosen as we have decimal numbers.

In [137]:
#Check data types
data.dtypes

Country     category
Features    category
Region      category
1980         float64
1981         float64
1982         float64
1983         float64
1984         float64
1985         float64
1986         float64
1987         float64
1988         float64
1989         float64
1990         float64
1991         float64
1992         float64
1993         float64
1994         float64
1995         float64
1996         float64
1997         float64
1998         float64
1999         float64
2000         float64
2001         float64
2002         float64
2003         float64
2004         float64
2005         float64
2006         float64
2007         float64
2008         float64
2009         float64
2010         float64
2011         float64
2012         float64
2013         float64
2014         float64
2015         float64
2016         float64
2017         float64
2018         float64
2019         float64
2020         float64
2021         float64
dtype: object



* To explore the dataset more accurately we can sample from rows to make sure there is no special case.


In [38]:
#Take sample from rows
data.sample(n=15)

Country              Features                   Region  \
64                     Moldova        net generation                  Eurasia   
370                      Nauru       net consumption           Asia & Oceania   
1213                Kyrgyzstan   installed capacity                   Eurasia   
1415                Mozambique  distribution losses                    Africa   
1182                Mauritania   installed capacity                    Africa   
1225                   Croatia   installed capacity                    Europe   
1036            American Samoa          net imports            Asia & Oceania   
1301                 Singapore   installed capacity            Asia & Oceania   
1187                     Niger   installed capacity                    Africa   
1581                   Ecuador  distribution losses   Central & South America   
273                    Senegal       net consumption                   Africa   
873           U.S. Territories              exports             North America   
61                     Georgia        net generation                  Eurasia   
520            Former U.S.S.R.              imports                   Eurasia   
1381                    Angola  distribution losses                    Africa   

         1980     1981     1982     1983     1984     1985     1986  ...  \
64        NaN      NaN      NaN      NaN      NaN      NaN      NaN  ...   
370   0.02232  0.02232  0.02232  0.02325  0.02418  0.02418  0.02511  ...   
1213      NaN      NaN      NaN      NaN      NaN      NaN      NaN  ...   
1415  0.18100  0.18200  0.03500  0.05200  0.06300  0.13600  0.05900  ...   
1182  0.06900  0.06900  0.06900  0.10500  0.10500  0.10500  0.10500  ...   
1225      NaN      NaN      NaN      NaN      NaN      NaN      NaN  ...   
1036  0.00000  0.00000  0.00000  0.00000  0.00000  0.00000  0.00000  ...   
1301  2.01000  2.10600  2.69100  2.69100  2.69100  2.74100  3.37100  ...   
1187  0.03700  0.05000  0.06300  0.06300  0.06300  0.06300  0.06300  ...   
1581  0.49900  0.53500  0.74400  0.77900  0.89500  0.95400  1.12800  ...   
273   0.49400  0.52500  0.52900  0.50500  0.53400  0.57900  0.58800  ...   
873       NaN      NaN      NaN      NaN      NaN      NaN      NaN  ...   
61        NaN      NaN      NaN      NaN      NaN      NaN      NaN  ...   
520   0.30000  0.30000  0.30000  0.30000  0.30000  1.90000  1.60000  ...   
1381  0.16900  0.16900  0.18000  0.19600  0.16800  0.20200  0.20200  ...   

          2012      2013       2014       2015       2016       2017  \
64     5.79900   4.50800   5.378000   6.049000   5.852000   4.962000   
370    0.01680   0.01570   0.016200   0.017593   0.019046   0.023749   
1213   3.86000   3.86200   3.865000   3.891000   3.812000   3.824000   
1415   2.24300   2.64700   2.612000   2.596000   2.492000   2.603000   
1182   0.22640   0.28100   0.290151   0.494000   0.511000   0.590000   
1225   4.23500   4.33100   4.442000   4.833500   4.923500   5.068700   
1036   0.00000   0.00000   0.000000   0.000000   0.000000   0.000000   
1301  10.48720  11.46850  13.137100  13.277100  13.407000  13.577100   
1187   0.13800   0.19200   0.230000   0.277000   0.288000   0.290000   
1581   3.31200   3.06000   3.165000   3.344000   3.499000   3.516000   
273    2.89048   3.01644   3.356560   3.583640   3.861140   4.133364   
873        NaN       NaN        NaN        NaN        NaN        NaN   
61     9.47468   9.86872  10.165840  10.605260  11.346840  11.305020   
520        NaN       NaN        NaN        NaN        NaN        NaN   
1381   0.69900   0.92600   1.068000   1.078000   1.167000   1.207373   

           2018       2019       2020       2021  
64     5.388240   5.695990   6.181000   6.656800  
370    0.026180   0.031886   0.034500   0.037426  
1213   3.824000   3.824000   4.626000   4.626000  
1415   2.663000   2.768000   2.768000   2.746750  
1182   0.620000   0.658000   0.656000   0.656000  
1225   5.127400   4.845000   4.738600   4.949600  



> As it is seen in the above sample, missing values are represented by different placeholders, such as '--', NaN, or empty cells. Handling this kind of data requires identifying and treating each type appropriately. Here's how we can recognize and handle these cases:



In [138]:
#Replace any non-numeric values with NaN
data.iloc[:,3:45].replace(to_replace=r'[^0-9]', value=np.nan, regex=True, inplace=True)

In the above approach:
*  The replace() function is used with a regular expression
(regex=True) to replace any non-numeric characters with NaN.
*  The regular expression r'[^0-9]' means anything that is not a digit will be replaced.

Now we can re-run the sum of null cells code.

In [139]:
#Check for the count of missing values
data.isnull().sum()

Country       0
Features      0
Region        0
1980        252
1981        252
1982        252
1983        252
1984        252
1985        252
1986        245
1987        252
1988        252
1989        252
1990        245
1991        252
1992        126
1993        119
1994        112
1995        112
1996        112
1997        105
1998        105
1999        105
2000        105
2001        105
2002        105
2003         98
2004         98
2005         98
2006         91
2007         91
2008         84
2009         84
2010         84
2011         84
2012         89
2013         89
2014         89
2015         89
2016         89
2017         89
2018         89
2019         89
2020         99
2021        101
dtype: int64



> It is obvious that the number of missing values increased.





*   To deal with missing values first we check if there exist rows in the dataset that contain only NaN values or a combination of NaN and zero values (except the first 3 columns).


In [140]:
#Check if all values in a row are NaN
data.iloc[:,3:45][data.iloc[:,3:45].isna().all(axis=1)]

1980  1981  1982  1983  1984  1985  1986  1987  1988  1989  ...  2012  \
138    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
146    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
159    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
183    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
368    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
376    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
389    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
413    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
598    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
606    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
619    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
643    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
828    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
836    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
849    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
873    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1058   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1066   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1079   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1103   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1288   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1296   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1309   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1333   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1518   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1526   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1539   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
1563   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   

      2013  2014  2015  2016  2017  2018  2019  2020  2021  
138    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
146    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
159    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
183    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
368    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
376    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
389    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
413    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
598    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
606    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
619    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
643    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
828    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
836    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
849    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
873    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1058   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1066   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1079   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1103   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1288   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1296   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1309   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1333   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1518   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1526   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1539   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1563  



> There are 28 rows in the dataset which all values (except the first 3 columns) are NaN. The proper approach is to delete these rows.



In [141]:
#Specify the range of columns
columns_to_check = data.columns[3:45]

#Drop rows where all values from column 3 to column 45 are NaN
data=data.dropna(subset=columns_to_check ,how='all')

In [142]:
#Check if all values in a row are either NaN or zero
data.iloc[:,3:45][(data.iloc[:,3:45].isna() | (data.iloc[:,3:45] == 0)).all(axis=1)]

1980  1981  1982  1983  1984  1985  1986  1987  1988  1989  ...  2012  \
128    0.0   0.0   0.0   0.0   0.0   0.0   0.0   NaN   NaN   NaN  ...   NaN   
163    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
358    0.0   0.0   0.0   0.0   0.0   0.0   0.0   NaN   NaN   NaN  ...   NaN   
393    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
461    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
1147   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1278   0.0   0.0   0.0   0.0   0.0   0.0   0.0   NaN   NaN   NaN  ...   NaN   
1313   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1508   0.0   0.0   0.0   0.0   0.0   0.0   0.0   NaN   NaN   NaN  ...   NaN   
1543   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

      2013  2014  2015  2016  2017  2018  2019  2020  2021  
128    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
163    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
358    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
393    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
461    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
1147   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1278   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1313   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1508   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1543   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[292 rows x 42 columns]



> There are 292 rows in the dataset which all values (except the first 3 columns) are NaN or zero. The proper approach is to delete these rows.



In [143]:
#Identify rows where all values from column 3 to column 45 are either NaN or zero
rows_to_drop = data[(data[columns_to_check].isna() | (data[columns_to_check] == 0)).all(axis=1)].index

#Drop the identified rows
data=data.drop(rows_to_drop)

In [144]:
data.shape

(1290, 45)

In [145]:
#Check for the count of missing values
data.isnull().sum()

Country       0
Features      0
Region        0
1980        211
1981        211
1982        211
1983        211
1984        211
1985        211
1986        207
1987        207
1988        207
1989        207
1990        200
1991        207
1992         81
1993         74
1994         70
1995         70
1996         70
1997         64
1998         64
1999         64
2000         64
2001         64
2002         64
2003         60
2004         60
2005         60
2006         53
2007         53
2008         46
2009         46
2010         46
2011         46
2012         54
2013         54
2014         54
2015         54
2016         54
2017         54
2018         54
2019         54
2020         54
2021         54
dtype: int64



> Based on the exploration above, it is evident that there are missing values in most of the columns.
> There are different approaches to handle missing values such as imputation, removal of rows or columns with missing values, or employing advanced techniques like machine learning algorithms to predict and fill in the missing data. The choice of method depends on the nature of the data and the specific goals of the analysis.

> Examining the pattern of missing values, we observe that in certain countries, time series values are recorded starting from one year onward, while in others, they are not recorded at all from the beginning. This kind of missing value is called time-series cross-section (TSCS) missing values. There is an article "What to Do about Missing Values in Time-Series Cross-Section Data" written by James Honaker and Gary King which is about handling with this kind of missing values but the package that contains this method in in R ("Amelia II"). While there isn't an exact equivalent to the Amelia II package in R, we can perform similar tasks using Python libraries.





In [146]:
pip install fancyimpute


In [109]:
pip install scikit-learn

In [147]:
from fancyimpute import IterativeImputer

# Create an imputer object
imputer = IterativeImputer()

# Fit the imputer on data and transform it
data_imputed=pd.DataFrame(imputer.fit_transform(data.iloc[:,3:45]), columns=columns_to_check)

#data = pd.concat([data.iloc[:, :3], data_imputed], axis=1)

/usr/local/lib/python3.10/dist-packages/sklearn/impute/_iterative.py:785: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.
  warnings.warn(


In [159]:
#Check for the count of missing values
data_imputed.isnull().sum()

1980    0
1981    0
1982    0
1983    0
1984    0
1985    0
1986    0
1987    0
1988    0
1989    0
1990    0
1991    0
1992    0
1993    0
1994    0
1995    0
1996    0
1997    0
1998    0
1999    0
2000    0
2001    0
2002    0
2003    0
2004    0
2005    0
2006    0
2007    0
2008    0
2009    0
2010    0
2011    0
2012    0
2013    0
2014    0
2015    0
2016    0
2017    0
2018    0
2019    0
2020    0
2021    0
dtype: int64

In [155]:
data.iloc[:, 3:45] = data_imputed



> The missing values are handled. Next step we check for duplicated data.



In [160]:
# Check for duplicated rows
data[data.duplicated()]

Empty DataFrame
Columns: [Country, Features, Region, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021]
Index: []

[0 rows x 45 columns]

In [130]:
data.to_excel("Global Electricity Statistics-cleaned.xlsx", index=False)
